packages/core/src/codewhisperer/util/zipUtil.ts

```diff
@@ -0,0 +1,129 @@
+    }
+
+    public getProjectName(uri: vscode.Uri) {
+        const projectPath = this.getProjectPath(uri)
+        return path.basename(projectPath)
+    }
+
+    public getProjectPath(uri: vscode.Uri) {
+        const workspaceFolder = vscode.workspace.getWorkspaceFolder(uri)
+        if (workspaceFolder === undefined) {
+            return this.getBaseDirPath(uri)
+        }
+        return workspaceFolder.uri.fsPath
+    }
+
+    protected getBaseDirPath(uri: vscode.Uri) {
+        return path.dirname(uri.fsPath)
+    }
+
+    protected async getTextContent(uri: vscode.Uri) {
+        const document = await vscode.workspace.openTextDocument(uri)
+        const content = document.getText()
+        return content
+    }
+
+    public reachSizeLimit(size: number): boolean {
+        return size > this.getFileScanPayloadSizeLimitInBytes()
+    }
+
+    protected async zipFile(uri: vscode.Uri) {
+        const zip = new admZip()
+
+        const projectName = this.getProjectName(uri)
+        const relativePath = vscode.workspace.asRelativePath(uri)
+
+        const content = await this.getTextContent(uri)
+
+        zip.addFile(path.join(projectName, relativePath), Buffer.from(content, 'utf-8'))
+
+        this._pickedSourceFiles.add(relativePath)
+        this._totalSize += statSync(uri.fsPath).size
+        this._totalLines += content.split(ZipConstants.newlineRegex).length
+
+        if (this.reachSizeLimit(this._totalSize)) {
+            getLogger().error(`Payload size limit reached.`)
+            throw new Error('Payload size limit reached.')
+        }
+
+        const zipFilePath = this.getZipDirPath() + CodeWhispererConstants.codeScanZipExt
+        zip.writeZip(zipFilePath)
+        return zipFilePath
+    }
+
+    protected getZipDirPath(): string {
+        if (this._zipDir === '') {
+            this._zipDir = path.join(
+                this._tmpDir,
+                CodeWhispererConstants.codeScanTruncDirPrefix + '_' + Date.now().toString()
+            )
+        }
+        return this._zipDir
+    }
+
+    public async generateZip(uri: vscode.Uri): Promise<ZipMetadata> {
+        try {
+            const zipDirPath = this.getZipDirPath()
+            const zipFilePath = await this.zipFile(uri)
+            getLogger().debug(`Picked source files: [${[...this._pickedSourceFiles].join(', ')}]`)
+            const zipFileSize = statSync(zipFilePath).size
+            return {
+                rootDir: zipDirPath,
+                zipFilePath: zipFilePath,
+                srcPayloadSizeInBytes: this._totalSize,
+                scannedFiles: new Set(this._pickedSourceFiles),
+                zipFileSizeInBytes: zipFileSize,
+                buildPayloadSizeInBytes: 0,
+                lines: this._totalLines,
+            }
+        } catch (error) {
+            getLogger().error('Zip error caused by:', error)
+            throw error
+        }
+    }
+
+    public async removeTmpFiles(zipMetadata: ZipMetadata) {
```

Comment:
Have you considered using a "streaming" approach for zipping files? Then temp files wouldn't be needed. That also avoids memory limits. And this would be useful for several other features in the codebase (featureDev, etc.).

- https://stackoverflow.com/a/71279922/152142
- https://github.com/archiverjs/node-zip-stream

Reply:
Will look into this in a separate PR